### PR TITLE
refactor deleteMeta and create mysql test

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

### Description:

I added remaining test of `internal/db/rdb/mysql.go`.
and, I refactored internal function `deleteMeta()` for easy to call from `DeleteMeta()` or `DeleteMetas()`.
Also, due to `dbr.Eq()` converts `IN` method when val is slice, I remove loop at `removeIPs()`.
ref: https://github.com/gocraft/dbr/blob/master/condition.go#L53

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/errors/mysql_test.go [hunk 1]
```diff
@@ -39,7 +42,7 @@ func TestErrMySQLNotFoundIdentity_Error(t *testing.T) {
 	}
 	defaultCheckFunc := func(w want, got string) error {
 		if !reflect.DeepEqual(got, w.want) {
-			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+			return errors.Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```
Comment (hlts2) [inline internal/errors/mysql_test.go:45]:
```suggestion
			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```

File: internal/errors/mysql_test.go [hunk 2]
```diff
@@ -112,7 +190,7 @@ func TestIsErrMySQLNotFound(t *testing.T) {
 	}
 	defaultCheckFunc := func(w want, got bool) error {
 		if !reflect.DeepEqual(got, w.want) {
-			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+			return errors.Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```
Comment (hlts2) [inline internal/errors/mysql_test.go:193]:
```suggestion
			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```

File: internal/errors/mysql_test.go [hunk 3]
```diff
@@ -182,7 +261,7 @@ func TestErrMySQLInvalidArgumentIdentity_Error(t *testing.T) {
 	}
 	defaultCheckFunc := func(w want, got string) error {
 		if !reflect.DeepEqual(got, w.want) {
-			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+			return errors.Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```
Comment (hlts2) [inline internal/errors/mysql_test.go:264]:
```suggestion
			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```

File: internal/errors/mysql_test.go [hunk 4]
```diff
@@ -255,7 +409,7 @@ func TestIsErrMySQLInvalidArgument(t *testing.T) {
 	}
 	defaultCheckFunc := func(w want, got bool) error {
 		if !reflect.DeepEqual(got, w.want) {
-			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
+			return errors.Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```
Comment (hlts2) [inline internal/errors/mysql_test.go:412]:
```suggestion
			return Errorf("got: \"%#v\",\n\t\t\t\twant: \"%#v\"", got, w.want)
```

**Review by hlts2 — CHANGES_REQUESTED** [review]

File: internal/db/rdb/mysql/mysql_test.go [hunk 5]
```diff
@@ -2846,84 +2828,237 @@ func Test_mySQLClient_deleteMetaWithTx(t *testing.T) {
...
-		*/
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
+			err := errors.ErrMySQLConnectionClosed
+			return test{
+				name: "return error when MySQL connection is closed",
+				args: args{
+					ctx: ctx,
+					val: "vald-01",
+				},
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:2839]:
```suggestion
			err := errors.ErrMySQLConnectionClosed
			return test{
				name: "return error when MySQL connection is closed",
				args: args{
					ctx: context.Background(),
					val: "vald-01",
				},
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 6]
```diff
@@ -2846,84 +2828,237 @@ func Test_mySQLClient_deleteMetaWithTx(t *testing.T) {
...
+						v.Store(true)
+						return
+					}(),
+				},
+				want: want{
+					err: err,
+				},
+				afterFunc: func(args) {
+					cancel()
+				},
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:2879]:
```suggestion
		func() test {
			err := errors.New("Begin error")
			return test{
				name: "return error when session.Begin returns error",
				args: args{
					ctx: context.Background(),
					val: "vald-01",
				},
				fields: fields{
					session: &mock.MockSession{
						BeginFunc: func() (dbr.Tx, error) {
							return nil, err
						},
					},
					connected: func() (v atomic.Value) {
						v.Store(true)
						return
					}(),
				},
				want: want{
					err: err,
				},
```

File: internal/errors/mysql_test.go (hunk 1 above)
Comment (vankichi) [reply internal/errors/mysql_test.go:45]: need to fix template 😓

File: internal/db/rdb/mysql/mysql_test.go [hunk 7]
```diff
@@ -2846,84 +2828,237 @@ func Test_mySQLClient_deleteMetaWithTx(t *testing.T) {
...
+			}
+		}(),
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:2883]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go [hunk 8]
```diff
@@ -2846,84 +2828,237 @@ func Test_mySQLClient_deleteMetaWithTx(t *testing.T) {
...
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:2911]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go (hunk 5 above)
Comment (vankichi) [reply internal/db/rdb/mysql/mysql_test.go:2839]: @hlts2 could you please tell me why you think that?

File: internal/db/rdb/mysql/mysql_test.go [hunk 9]
```diff
@@ -2846,84 +2828,237 @@ func Test_mySQLClient_deleteMetaWithTx(t *testing.T) {
...
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:2962]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go [hunk 10]
```diff
@@ -3011,112 +3112,114 @@ func Test_mySQLClient_DeleteMeta(t *testing.T) {
...
-			if test.beforeFunc != nil {
-				test.beforeFunc(test.args)
-			}
-			if test.afterFunc != nil {
-				defer test.afterFunc(test.args)
-			}
-			if test.checkFunc == nil {
-				test.checkFunc = defaultCheckFunc
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:3116]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go [hunk 11]
```diff
@@ -3174,82 +3260,104 @@ func Test_mySQLClient_DeleteMetas(t *testing.T) {
...
-		           eventReceiver: nil,
-		           dbr: nil,
-		           },
-		           want: want{},
-		           checkFunc: defaultCheckFunc,
-		       }
-		   }(),
-		*/
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:3264]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go [hunk 12]
```diff
@@ -3174,82 +3260,104 @@ func Test_mySQLClient_DeleteMetas(t *testing.T) {
...
+					},
+				},
+				want: want{},
+				afterFunc: func(args) {
+					cancel()
+				},
+			}
+		}(),
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:3309]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go [hunk 13]
```diff
@@ -3338,114 +3412,360 @@ func Test_mySQLClient_SetIPs(t *testing.T) {
...
-				dialerFunc:           test.fields.dialerFunc,
-				tlsConfig:            test.fields.tlsConfig,
-				maxOpenConns:         test.fields.maxOpenConns,
-				maxIdleConns:         test.fields.maxIdleConns,
-				session:              test.fields.session,
-				connected:            test.fields.connected,
-				eventReceiver:        test.fields.eventReceiver,
-				dbr:                  test.fields.dbr,
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:3416]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go [hunk 14]
```diff
@@ -3338,114 +3412,360 @@ func Test_mySQLClient_SetIPs(t *testing.T) {
...
+						v.Store(false)
+						return
+					}(),
+				},
+				want: want{
+					err: err,
+				},
+				afterFunc: func(args) {
+					cancel()
+				},
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:3439]:
```suggestion
```

File: internal/db/rdb/mysql/mysql_test.go [hunk 15]
```diff
@@ -3338,114 +3412,360 @@ func Test_mySQLClient_SetIPs(t *testing.T) {
...
+			}
+		}(),
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:3443]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

File: internal/db/rdb/mysql/mysql_test.go [hunk 16]
```diff
@@ -3503,82 +3806,167 @@ func Test_mySQLClient_RemoveIPs(t *testing.T) {
...
-		           eventReceiver: nil,
-		           dbr: nil,
-		           },
-		           want: want{},
-		           checkFunc: defaultCheckFunc,
-		       }
-		   }(),
-		*/
+		func() test {
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (hlts2) [inline internal/db/rdb/mysql/mysql_test.go:3810]: same
https://github.com/vdaas/vald/pull/725/files#r495819458

**Review by hlts2 — CHANGES_REQUESTED** [review]

File: internal/db/rdb/mysql/mysql_test.go (hunk 5 above)
Comment (hlts2) [reply internal/db/rdb/mysql/mysql_test.go:2839]: I didn't handle context.Done in this function, so I think it wasn't necessary :thinking:

**Review by kevindiu — APPROVED** [review]
LGTM

**Review by hlts2 — APPROVED** [review]
LGTM

**Pull request merged**